src/main/scala/is/hail/utils/richUtils/RichIterableGenotype.scala

```diff
@@ -0,0 +1,8 @@
+package is.hail.utils.richUtils
+
+import is.hail.utils.{IntIterator, HardCallIterator}
+import is.hail.variant.Genotype
+
+class RichIterableGenotype(it: Iterable[Genotype]) {
+  def hardCallIterator: HardCallIterator = new HardCallIterator(it.iterator)
```

Comment:
With this definition, you never use your new code.  You have to match on the type of it here.

I'd probably define `HardCallIterator` inline since it is (should be) only used once.  This should return an `IntInterator`.

The goal here (as with the `MutableGenotype` stuff) is eliminate all per-Genotype allocations.